### ISSUE TITLE
[java] EmptyMethodInAbstractClassShouldBeAbstract: False positive for final empty methods

### ISSUE THREAD

**Issue by Vampire** [pmd/pmd#6279]
**Affects PMD Version:** 7.18.0

**Rule:** [EmptyMethodInAbstractClassShouldBeAbstract](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#emptymethodinabstractclassshouldbeabstract)

**Description:**

When overriding a method and making it final in an abstract class, this rule should not give a finding.
This is for example used to prevent a finalizer attack by overriding the `finalize` method and making the method `final` without having an implementation in the method body.

**Code Sample demonstrating the issue:**

```java
package example;

public abstract class PMDTest {
    public PMDTest() {
        throw new RuntimeException();
    }

    @Override
    protected final void finalize() {
        // prevent finalizer attack
    }
}
```

**Expected outcome:**

PMD reports a violation but that's wrong. That's a false positive.

**Running PMD through:** *Gradle*

**Comment by mrclmh** [comment]
Thanks for the report. Example was missing `abstract`:

```Java
package example;

public abstract class PMDTest {
    public PMDTest() {
        throw new RuntimeException();
    }

    @Override
    protected final void finalize() {  // violation
        // prevent finalizer attack
    }
}
```

**Comment by Vampire** [comment]
No, the example is not missing `abstract`.
Whether the class is `abstract` or not should be irrelevant.

**Comment by mrclmh** [comment]
I tested it again. For me the rule [EmptyMethodInAbstractClassShouldBeAbstract](https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#emptymethodinabstractclassshouldbeabstract) only reports a violation, when the class in the example is abstract.
As far as I understand the XPath expression in the docs, the rule should only check for abstract classes.

Do you get a report for this rule, when the class in the example is not abstract?

**Comment by Vampire** [comment]
Ah, sorry, I'm stupid.
Even the name of the rule says `...InAbstractClass...`.
I added the missing `abstract` to the example in the OP.

**Comment by mrclmh** [comment]
No worries. We've all been there.
